cdap-ui/app/cdap/components/PreviewData/utilities.ts

```diff
@@ -118,57 +128,59 @@ export function fetchPreview(
   );
 }
 
-export function getPreviousStageName(selectedStage: string, adjMap: IAdjacencyMap) {
+export function getPreviousStageNames(selectedStage: string, adjMap: IAdjacencyMap) {
   const keys = adjMap.keySeq().toArray();
-  const prevStageName = keys.find((key) => adjMap.get(key).includes(selectedStage));
-  return prevStageName;
-}
-
-function getPreviousStage(previousStageName: string, stages: IStage[]) {
-  return stages.find((stage) => stage.name === previousStageName);
+  const prevStageNames = keys.filter((key) => adjMap.get(key).includes(selectedStage));
+  return prevStageNames;
 }
 
 function getTracersAndPreviousStageInfo(
-  selectedStage: string,
+  selectedStageName: string,
   adjacencyMap: IAdjacencyMap,
   stages: IStage[],
   connections: IConnection[]
 ) {
-  let tracers: List<string> = List([selectedStage]);
-  let previousStageName = getPreviousStageName(selectedStage, adjacencyMap);
+  let tracers: List<string> = List([selectedStageName]);
+  const previousStageNames: string[] = getPreviousStageNames(selectedStageName, adjacencyMap);
 
   // source nodes have no previous stage
-  if (!previousStageName) {
+  if (isEmpty(previousStageNames)) {
     return { tracers: tracers.toArray() };
   }
 
-  let previousStage = getPreviousStage(previousStageName, stages);
-  let previousStagePort;
-
-  if (previousStage.plugin.type === 'splittertransform') {
-    const previousStageConnection = connections.find(
-      (connection) => connection.from === previousStageName && connection.to === selectedStage
-    );
-    if (previousStageConnection) {
-      previousStagePort = previousStageConnection.port;
-    }
-  } else {
-    // If we have multiple condition nodes in a row, we have traverse back
-    // until we find a node that actually has records out
-    while (previousStage && previousStage.plugin.type === 'condition') {
-      previousStageName = getPreviousStageName(previousStageName, adjacencyMap);
-      previousStage = stages.find((stage) => stage.name === previousStageName);
+  const previousStages: IPreviousStageInfo = {};
+  previousStageNames.forEach((previousStageName) => {
+    let previousStage = stages.find((stage) => stage.name === previousStageName);
+    previousStages[previousStageName] = {};
+
+    if (previousStage.plugin.type === 'splittertransform') {
+      const previousStageConnection = connections.find(
+        (connection) => connection.from === previousStageName && connection.to === selectedStageName
+      );
+      if (previousStageConnection) {
+        previousStages[previousStageName].port = previousStageConnection.port;
+      }
+    } else {
+      // If we have multiple condition nodes in a row, we have traverse back
+      // until we find a node that actually has records out
+      while (previousStage && previousStage.plugin.type === 'condition') {
```

Comment:
Are there any unusual possibilities to consider here? E.g. is it possible to have no nodes that meet the condition?

Reply:
Most plugins will be covered with getPreviousStageNames, since the previous stage name is usually just the plugin(s) that are just before the selected. plugin. The only exception is when the previous stage is a conditional, which may or may not have output, so we continue down the pipeline to get the output of the next closest non-conditional plugin. 

I am honestly not sure why we need to get the port of the splitter plugins, because we don't use that information anywhere currently. I have added a comment to either remove or make sure we actually use the port info before the next release (6.1.4).